### PR TITLE
add direction for saveat + test

### PR DESCRIPTION
**Pull request by lockwo**
Addresses https://github.com/patrick-kidger/diffrax/issues/425

### CODE REVIEW

**Comment by patrick-kidger** [comment]
Awesome -- thanks for this!
I can see there's still some failing tests. I've not dug into them, but please let me know once this is ready for review :)

**Comment by lockwo** [comment]
It seems as though that test fails on main. I was very confused by it (since I tried hard to make sure there were no closure issues) and when I just clone diffrax main and run `python -m pytest test/test_adjoint.py -k 'test_implicit'` I see the same error: `FAILED test/test_adjoint.py::test_implicit - AssertionError`

**Comment by lockwo** [comment]
So you can also see (without just taking my work for it), this PR, which changes nothing as of yet (random 1 line README change), still fails tests: https://github.com/patrick-kidger/diffrax/pull/430.

**Comment by patrick-kidger** [comment]
Fixed in #431! I'm not sure why this ever passed...
We'll get that merged in and then rebase this PR on top of the new main.

File: diffrax/_adjoint.py [hunk 1]
```diff
@@ -482,7 +484,16 @@ def loop(
...
-        inputs = (args, terms, self, kwargs, solver, saveat, init_state)
+        inputs = (
+            args,
+            terms,
+            self,
+            kwargs,
+            solver,
+            saveat,
+            init_state,
+            kwargs.pop("direction"),
```
Comment (patrick-kidger) [inline diffrax/_adjoint.py:495]: I think it'd be better to just make `direction` be another keyword argument to `loop`.
(Or for that matter could it just be left in `kwargs` and passed through? So no code changes here whatsoever?)

File: diffrax/_integrate.py [hunk 2]
```diff
@@ -405,7 +406,7 @@ def _body_fun(_save_state):
                 _ts = _save_state.ts.at[_save_state.save_index].set(_t)
                 _ys = jtu.tree_map(
                     lambda __y, __ys: __ys.at[_save_state.save_index].set(__y),
-                    fn(_t, _y, args),
+                    fn(direction * _t, _y, args),
```
Comment (patrick-kidger) [inline diffrax/_integrate.py:409]: I think this change needs to include in all the other places we call `fn`, i.e. when saving at `t0`, `t1`, `steps`, dense`.

I think it'd better to simply wrap `.fn` inside of `saveat`, during the initial set-up for the solve. It's a bit of a mouthful, but something like this:
```python
def _subsaveat_direction_fn(x):
    if _is_subsaveat(x):
        direction_fn = lambda t, y, args: x.fn(direction * t, y, args)
        return eqx.tree_at(lambda x: x.fn, x, direction_fn)
    else:
        return x

saveat = jtu.tree_map(_subsaveat_direction, saveat, is_leaf=_is_subsaveat)
```

File: test/test_integrate.py [hunk 3]
```diff
@@ -398,6 +398,30 @@ def g(t, y, args):
             assert tree_allclose(sol1.derivative(ti), -sol2.derivative(-ti))
 
 
+def test_reverse_time_save():
```
Comment (patrick-kidger) [inline test/test_integrate.py:401]: As above I think it's worth including other kinds of `saveat` here, which include `t0`, `t1` etc, to be sure that these cases don't slip through.

**Review by patrick-kidger — COMMENTED** [review]
Great stuff! I've just merged the test fix, so if you'd like to rebase on top of that then that would be great.

File: diffrax/_integrate.py (hunk 2 above)
Comment (lockwo) [reply diffrax/_integrate.py:409]: That's actually what I had originally, but I thought direction would then get closured in and cause issues. I will roll back to that version

File: test/test_integrate.py (hunk 3 above)
Comment (lockwo) [reply test/test_integrate.py]: Expanded to include all the flags of saveat. Note that I specify max steps because if that is greater and steps is being logged, it multiples the jnp.inf by direction at the end which results in a mismatch. I don't see a great way to get the custom function to match that and given that those infinities are just padding/holding space, I don't consider it an actual issue

File: diffrax/_adjoint.py (hunk 1 above)
Comment (lockwo) [reply diffrax/_adjoint.py]: Moved to avoid direction as keyword altogether with just modifying the function

File: test/test_integrate.py (hunk 3 above)
Comment (patrick-kidger) [reply test/test_integrate.py]: Hmm, that surprises me. The `jnp.inf` should just be the initial buffer values, I don't think they're touched afterwards. In particular I don't think the choice of `direction` should affect them?

Other than that this PR LGTM!

Comment (lockwo) [reply test/test_integrate.py]: Sepcifically, the output ys vs ts are [3.9 3.8 3.7 3.6 3.5 3.4 3.3 3.2 3.1 3.  2.9 2.8 2.7 2.6 2.5 2.4 2.3 2.2
 2.1 2.  1.9 1.8 1.7 1.6 1.5 1.4 1.3 1.2 1.1 1.  0.9 0.8 0.7 0.6 0.5 0.4
 0.3 inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf
 inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf
 inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf
 inf inf inf inf inf inf inf inf inf inf]

vs

 [ 3.9  3.8  3.7  3.6  3.5  3.4  3.3  3.2  3.1  3.   2.9  2.8  2.7  2.6
  2.5  2.4  2.3  2.2  2.1  2.   1.9  1.8  1.7  1.6  1.5  1.4  1.3  1.2
  1.1  1.   0.9  0.8  0.7  0.6  0.5  0.4  0.3 -inf -inf -inf -inf -inf
 -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf
 -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf
 -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf
 -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf
 -inf -inf]

The ts get a direction multiplied by them at the end, but it is initialized to `inf` without getting direction multiplied. I think this might be a bug, so I fixed it as well and now all the ts inf get initialized with respect to the direction (        ts = jnp.full(out_size, direction * jnp.inf, dtype=time_dtype)). Maybe not a bug since these `inf` are technically meaningless, but at least now the tests pass no problem and I can remove max steps.

Comment (lockwo) [reply test/test_integrate.py]: Hmm, I guess that actually breaks `assert tree_allclose(sol1.ts, -cast(Array, sol2.ts), equal_nan=True)` for steps=True because then the infs are not opposite signs. Not sure if they should be or not, since they are basically superfluous. I will defer to whatever you think. I can revert the last commit and keep max steps and the infs will be sign flipped, or I can make a change to what this failing test expects and the infs will always be positive.

Comment (patrick-kidger) [reply test/test_integrate.py]: Right. So I think this is indicative of a totally separate bug: in some cases we've been returning `-inf` all along:
```python
diffrax.diffeqsolve(diffrax.ODETerm(lambda t, y, args: -y), diffrax.Tsit5(), 1, 0, -0.1, 1, saveat=diffrax.SaveAt(steps=True), max_steps=50).ts
```
This shouldn't be the case. I think we should always return `+inf` for our padding!

It looks like the culprit is as you've identified, namely that we flip things here:

https://github.com/patrick-kidger/diffrax/blob/c6cc85c92b0e14e24092830584d0397a1990d091/diffrax/_integrate.py#L1018-L1020

I think we should adjust the failing test you've highlighted, so that the infs are always positive.

Comment (lockwo) [reply test/test_integrate.py]: Great, I changed that test, and all padding should be positive now.

File: diffrax/_integrate.py [hunk 4]
```diff
@@ -881,6 +881,18 @@ def _check_subsaveat_ts(ts):
 
     saveat = eqx.tree_at(_get_subsaveat_ts, saveat, replace_fn=_check_subsaveat_ts)
 
+    def _subsaveat_direction_fn(x):
+        if _is_subsaveat(x):
+            if x.fn is not save_y:
```
Comment (patrick-kidger) [inline diffrax/_integrate.py:886]: Why this exclusion? I don't it should actually matter? (I'd rather not have an extra branch unless we need it.)

**Review by patrick-kidger — COMMENTED** [review]
Ultranit aside this LGTM!

File: diffrax/_integrate.py (hunk 4 above)
Comment (patrick-kidger) [reply diffrax/_integrate.py:886]: Ah, makes sense!
Alright -- merged! :D
